### PR TITLE
Use pep440_rs and pep508_rs

### PR DESCRIPTION
**Pull request by konstin**
This switches from string heuristics to proper PEP 440 and PEP 508 parsing. This includes full support for `requires-python`, extra handling and wasmtime detection.

### CODE REVIEW

File: Cargo.lock [hunk 1]
```diff
@@ -1353,13 +1354,32 @@ source = "registry+https://github.com/rust-lang/crates.io-index"
...
 dependencies = [
  "lazy_static",
  "regex",
+ "serde",
+ "tracing",
+ "unicode-width",
+]
+
+[[package]]
+name = "pep508_rs"
```
Comment (konstin) [inline Cargo.lock:1370]: effectively this only adds only one new dependency

File: src/develop.rs [hunk 2]
```diff
@@ -84,12 +85,28 @@ pub fn develop(
...
-                    .replace(&format!(" and extra == '{extra}'"), "")
-                    .replace(&format!("; extra == '{extra}'"), "");
+                pkg.marker = pkg.marker.and_then(|marker| -> Option<MarkerTree> {
+                    match marker.clone() {
+                        MarkerTree::Expression(MarkerExpression {
+                            l_value: MarkerValue::Extra,
+                            operator: MarkerOperator::Equal,
+                            r_value: MarkerValue::QuotedString(extra_value),
+                        }) if &extra_value == extra => None,
+                        MarkerTree::And(and) => match &*and {
```
Comment (konstin) [inline src/develop.rs:97]: not sure how to deal with the borrow checker better here without vec patterns

File: src/metadata.rs [hunk 3]
```diff
@@ -61,6 +64,42 @@ pub struct Metadata21 {
     pub entry_points: IndexMap<String, IndexMap<String, String>>,
 }
 
+impl Metadata21 {
+    /// Initializes with name, version and otherwise the defaults
+    pub fn new(name: String, version: Version) -> Self {
```
Comment (konstin) [inline src/metadata.rs:69]: Since Version does not have a default

File: src/python_interpreter/mod.rs [hunk 4]
```diff
@@ -704,47 +717,49 @@ impl PythonInterpreter {
...
-            if matches!(bridge, BridgeModel::Cffi)
+            if *bridge == BridgeModel::Cffi
                 || bridge.is_bindings("pyo3")
                 || bridge.is_bindings("pyo3-ffi")
             {
                 executables.extend(
-                    (min_python_minor..=MAXIMUM_PYPY_MINOR).map(|minor| format!("pypy3.{minor}")),
+                    (min_python_minor..=MAXIMUM_PYPY_MINOR)
+                        .filter(|minor| {
+                            requires_python
```
Comment (konstin) [inline src/python_interpreter/mod.rs:755]: this is some duplication unfortunately

**Review by messense — APPROVED** [review]
LGTM, thanks!

bors r+

**Pull request merged**